Plugin.cs

```diff
@@ -108,23 +109,26 @@ private void BSEvents_levelSelected(LevelCollectionViewController arg1, IPreview
         {
             if (level is CustomPreviewBeatmapLevel customLevel)
             {
-                var songData = Collections.RetrieveExtraSongData(Hashing.GetCustomLevelHash(customLevel), customLevel.customLevelPath);
-                Collections.SaveExtraSongData();
-
-                if (songData == null)
-                {
-                    return;
-                }
-
-                if (CustomSongPlatforms && !string.IsNullOrWhiteSpace(songData._customEnvironmentName))
+                Task.Run(() =>
                 {
-                    Logging.Logger.Debug("Custom song with platform selected");
-                    CustomSongPlatformSelectionDidChange?.Invoke(true, songData._customEnvironmentName, songData._customEnvironmentHash, customLevel);
-                }
-                else
-                {
-                    CustomSongPlatformSelectionDidChange?.Invoke(false, songData._customEnvironmentName, songData._customEnvironmentHash, customLevel);
-                }
+                    var songData = Collections.RetrieveExtraSongData(Hashing.GetCustomLevelHash(customLevel), customLevel.customLevelPath);
+                    Collections.SaveExtraSongData();
+
+                    if (songData == null)
+                    {
+                        return;
+                    }
+
+                    if (CustomSongPlatforms && !string.IsNullOrWhiteSpace(songData._customEnvironmentName))
+                    {
+                        Logging.Logger.Debug("Custom song with platform selected");
+                        CustomSongPlatformSelectionDidChange?.Invoke(true, songData._customEnvironmentName, songData._customEnvironmentHash, customLevel);
```

Comment:
Probably not a good idea to invoke the events off-main thread. Only have the loading of the data be in the Task and leave others outside

Reply:
Yep was really tired yesterday and will probably just have the saving in the Task. I also plan on retrieving all the extra data on initial loading instead of doing this each time you select a song, to prevent having to save that file that many times and possibly corrupting the data.

Reply:
Yep was really tired yesterday and will probably just have the saving in the Task. I also plan on retrieving all the extra data on initial loading instead of doing this each time you select a song, to prevent having to save that file that many times and possibly corrupting the data.